### ISSUE TITLE
Libtorrent looking for wrong boost version on Debian Stretch

### ISSUE THREAD

**Issue by Protected** [arvidn/libtorrent#2584]
Sorry if this is the wrong place for this but I tried forums and IRC with little luck, and this is driving me nuts.

I have a server that was previously running Jessie and libtorrent (package libtorrent-rasterbar8) was functioning properly. After upgrading to Stretch, the new package (libtorrent-rasterbar9) does not work at all:

    server ~/libtorrent # python -c "import libtorrent; print libtorrent.version"
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
    ImportError: libboost_system.so.1.55.0: cannot open shared object file: No such file or directory
    server ~/libtorrent # locate libboost_system
    /usr/lib/x86_64-linux-gnu/libboost_system.a
    /usr/lib/x86_64-linux-gnu/libboost_system.so
    /usr/lib/x86_64-linux-gnu/libboost_system.so.1.62.0

I've tried to remove and reinstall all packages, and I also tried to compile manually using b2 and setup.py (after generating it with ./configure) but nothing worked. I'm using master. The error also showed libboost_chrono previously, but was otherwise exactly the same.

Is there a way to fix this manually? Also, could the package's dependencies on debian stretch be broken?

EDIT: Forgot to say, I also tried symbolic links with the old version filename; nothing changed.

**Comment by arvidn** [comment]
my suspicion is that the python module itself is stale, or not being rebuilt. It could also be the case that the rebuilt module is installed in a different location and the old location takes precedence.
did you try putting the python module you built with ``b2`` in the current directory when running python?

**Comment by Protected** [comment]
Thank you for the reply. I tried to purge the packages and reinstall but still no success. I confirmed the latest version of the packages that have two versions in the system is being used, for each package involved.

If I use b2 as described in the main readme I won't get the python bindings, right? I don't know how to use boost build to get the python bindings. I don't understand what the boost build dir even is (sorry!)

I tried the following:

    ./autotool.sh
    ./configure --enable-python-binding
    python setup.py build
    python setup.py install

At this point the error persists.

I then manually copied the libtorrent in build to the directory where I was running python and tried the test above; The error still persists, unfortunately.

Config results: https://pastebin.com/x0iw4YCJ

Install results: https://pastebin.com/mD18GBji

**Comment by Protected** [comment]
Figured out the copy of libtorrent in use was still

    /usr/lib/x86_64-linux-gnu/libtorrent-rasterbar.so.9

I replace it with a symbolic link to the newly compiled libtorrent. It's also looking for the 1.55 libboosts. Linking those names to their 1.62 counterparts yields:

    # python -c "import libtorrent; print libtorrent.version"
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
    ImportError: ./libtorrent.so: undefined symbol: _ZNK10libtorrent12file_storage10piece_sizeENS_3aux14strong_typedefIiNS1_15piece_index_tagEvEE

So right now I know I'm in the right place, but still need to figure out why all of the instances of libtorrent (linked manually or by the package manager) are referencing a bunch of libraries that no longer exist.

**Comment by Protected** [comment]
Tried boost-build. Installed, created the user-config with gcc and the correct version of python and ran bjam in the bindings folder as described in the documentation. I get these now:
    server ~ # python -c "import libtorrent; print libtorrent.version"
-c:1: RuntimeWarning: to-Python converter for boost::shared_ptr<libtorrent::alert> already registered; second conversion method ignored.
    1.1.1.0

**Comment by Protected** [comment]
I removed libtorrent and every boost package (and their dependencies), nuked the leftovers in the python modules folder and other related files I could get my hands on and then reinstalled from the repository, but am still getting the error above.

**Comment by arvidn** [comment]
it sounds like a package issue.
The safest way to make sure everything is compatible is to link everything statically. That requires having the boost source though (set the ``BOOST_ROOT`` environment variable to the path of the boost source directory).

build the python binding by cd into the libtorrent/bindings/python directory and run:

``b2 lintorrent-link=static boost-link=static stage_module``

That should drop a ``libtorrent.so`` in the current directory that's a python module.
There's still a possibility of having a python2/python3 mismatch, but it doesn't sound like you've hit that issue so far.

**Comment by Protected** [comment]
# b2 lintorrent-link=static boost-link=static stage_module
    BOOST_ROOT = /root/boost_1_65_1/
    CXXFLAGS =
    LDFLAGS =
    OS = LINUX
    /usr/local/share/boost-build/src/build/feature.jam:139: in feature.feature from module feature
    error: unknown attributes: hidden
    error: in feature declaration:
    error: feature "deduced-address-model" : "32" "64" : "propagated" "optional" "composite" "hidden"
    /root/boost_1_65_1/boostcpp.jam:624: in load from module boostcpp

**Comment by Protected** [comment]
Is there just a precompiled, portable libtorrent available for download somewhere? I find myself in need of 1.0.11 (for deluge compatibility). I have spent many, many hours on this that I could not afford to spend. I can't believe this is so difficult and troublesome. (Thanks for trying to help, though.)

**Comment by Dnkhatri** [comment]
I have been using the latest versions of libtorrent by adding the qbittorrent develop ppa https://launchpad.net/~qbittorrent-team/+archive/ubuntu/qbittorrent-unstable and updating libtorrent it works with no problems for deluge 1.3.15 for me though memory consumption seem to be on the high side with 3-400 torrents.

**Comment by arvidn** [comment]
@Protected that error suggests you use an older version of boost-build (installed on your system) instead of the one that come with boost. I suppose that does make things a bit more complicated.
Basically, point the ``BOOST_BUILD_PATH`` environment variable to the ``boost/tools/build/v2`` folder. And if the ``bjam`` executable is too old a version, you may have to build that too.

It's not really possible to provide pre-built *portable* binaries. At least not a single one, as there are so many combinations of (platform, architecture, compiler, runtime) combinations for that to be practical (for me to provide at least).

**Comment by Protected** [comment]
Right. I apologize for my frustration, which I'm sure is understandable.

I'd read those instructions before, but I have never been able to find a tools/build/v2 path anywhere during these proceedings.

**Comment by Protected** [comment]
Some more clarifications about boost-build: When I install it from the website, I get the version 2014-svn, and when I install it from tools/build (no v2) I get 2015-git, but also a few warnings about how the executable isn't the same version as the core.

**Comment by Protected** [comment]
Installed boost-build from 1.65 (to /usr/bin, etc), which is the latest possible version, and that seems to have fixed mismatched versions and warnings in boost-build itself. I then used it (b2 ...) to build boost 1.60, and that ran without any issues too. I copied the resulting libraries (stage/lib) to the library path. So I'm fully boost-ed right now. 

However, b2 on libtorrent 1.0.11 still stops immediately with a complaint about not knowing what "hidden" is about, no matter what arguments I use.

**Comment by arvidn** [comment]
Could you post the full error from ``b2``? I believe the problem there is that newer versions of b2 introduced a feature called "hidden", which clashes with the custom feature I defined in the Jamfile for the python binding. You can just comment out any mentions of ``hidden`` in the Jamfile to get it to build.

This was fixed in later versions of libtorrent. keep in mind that the 1.0.x series is several years old by now.

**Comment by aldenml** [comment]
messing up with the system paths are never a good idea. If what are you missing is `libboost_system.so.1.55.0`, aren't you able to install that specific version of boost, or build/install that specific version of boost, and leave the `libtorrent` system provided package as is?

**Comment by Protected** [comment]
So, currently I get a ton of warnings when running b2 stage_module on bindings/python, but it does produce a libtorrent.so, so I think that's fine. I tried to copy it to:

    /usr/lib/python2.7/dist-packages

And now when I run the command at the very top of this issue I get:

    ImportError: libtorrent.so.1.0.11: cannot open shared object file: No such file or directory

So it sees the module, but expects something else. I tried to build the library and copying it all over the place but python just won't see it. Still fooling around with this.

**Comment by Protected** [comment]
Turns out I can statically link libtorrent even if I can't statically link boost. Since boost was compiled and installed with no issue, I believe that will not be a problem.

    server ~/libtorrent-rasterbar-1.0.11/bindings/python # python -c "import libtorrent; print libtorrent.version"
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
    ImportError: ./libtorrent.so: undefined symbol: libiconv_open

**Comment by Protected** [comment]
Getting rid of iconv:

    server ~/libtorrent-rasterbar-1.0.11/bindings/python # python -c "import libtorrent; print libtorrent.version"
    1.0.11.0
    server ~/libtorrent-rasterbar-1.0.11/bindings/python #

**Comment by arvidn** [comment]
If you’re up for it, feel free to contribute a pull request with doc updates!

**Comment by Protected** [comment]
I'm not going to open another ticket yet because this is obviously old news, but any hints would be appreciated. When running deluged:

    $ Unhandled error in Deferred:

    assertion failed. Please file a bugreport at https://github.com/arvidn/libtorrent/issues
    Please include the following information:

    version: 1.0.11.0
    bbb38a9

    file: '../../src/natpmp.cpp'
    line: 200
    function: void libtorrent::natpmp::delete_mapping(int)
    expression: index < int(m_mappings.size()) && index >= 0


    stack:
    1:
    2:
    3:
    4:
    5:
    6:
    7:
    8:
    9:
    10:
    11:
    12: boost::asio::detail::task_io_service::do_run_one(boost::asio::detail::scoped_lock<boost::asio::detail::posix_mutex>&, boost::asio::detail::task_io_service_thread_info&, boost::system::error_code const&)
    13: boost::asio::detail::task_io_service::run(boost::system::error_code&)
    14: boost::asio::io_service::run(boost::system::error_code&)
    15:
    16:
    17:
    18:
    19:
    20: boost_asio_detail_posix_thread_function
    21:
    22: clone

Likewise on 1.0.9 .

**Comment by Protected** [comment]
Unfortunately, boost 1.55 did not get rid of the above issue.

**Comment by aldenml** [comment]
which one? the one in the first comment? or the assert in your last one?

**Comment by Protected** [comment]
The assert. 

But it seems I owe  you an apology. Adding the packages on top of my mess (I made sure the boost 1.55 stuff remains there and pointed at by the generic library names) I was able to run deluge and it hasn't crashed *yet* (I didn't really do anything with it, only tried torguard to make sure ltconfig is functioning, which is the one thing in this setup I really need). 

I was referencing such threads as:

http://forum.deluge-torrent.org/viewtopic.php?f=7&t=54715&p=227253  
http://forum.deluge-torrent.org/viewtopic.php?t=53939  
etc.

Yet libtorrent-rasterbar9 is definitely 1.1:

     python -c "import libtorrent; print libtorrent.version"
    -c:1: RuntimeWarning: to-Python converter for boost::shared_ptr<libtorrent::alert> already registered; second conversion method ignored.
    1.1.1.0

I guess I'll fool around a bit and update here about whether everything's working.

Is there any way to get rid of that warning?
